### PR TITLE
Add `http2` directive instead of `listen` option

### PR DESCRIPTION
**Pull request by C24-AK**
#### Pull Request (PR) description

This PR is meant to create the http2 directive: http://nginx.org/en/docs/http/ngx_http_v2_module.html#http2

#### This Pull Request (PR) fixes the following issues

Fixes following warning in nginx:
nginx: [warn] the "listen ... http2" directive is deprecated, use the "http2" directive instead.

### CODE REVIEW

**Comment by kenyon** [comment]
The tests need to be updated to match this change.

**Comment by C24-AK** [comment]
> The tests need to be updated to match this change.

Where can I adjust the nginx_version for the tests? Since 1.25.1 is required.

**Comment by kenyon** [comment]
You can specify an nginx version in an acceptance test's puppet code, just like in regular puppet code. Since the `http2` directive doesn't exist in older versions, I think you have to have some logic that only uses it if the nginx version is 1.25.1 or newer.

**Comment by C24-AK** [comment]
@kenyon Hey, can you tell me why archlinux is the only test failing? It seems like other Pull Requests have the same problem.

**Comment by kenyon** [comment]
@C24-AK Archlinux test failure is expected for now.

**Comment by C24-AK** [comment]
@kenyon Hey, what to do on following error?
This happens outside of examples:

> JSON::ParserError:
> unexpected token at ''
> ./vendor/bundle/ruby/2.7.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
> ./vendor/bundle/ruby/2.7.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:99:in `load_filters'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:80:in `block in merge_filters'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:79:in `each'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:79:in `merge_filters'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:201:in `run_report'
> ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:186:in `report!'
> ./vendor/bundle/ruby/2.7.0/gems/voxpupuli-test-7.0.2/lib/voxpupuli/test/spec_helper.rb:21:in `block (2 levels) in <top (required)>'

**Comment by kenyon** [comment]
Yeah I don't know what is causing that. Some debugging required.

**Comment by C24-AK** [comment]
Is this something I need to fix for this PR to be merged?

File: templates/server/server_ssl_header.erb [hunk 1]
```diff
@@ -4,10 +4,10 @@
 server {
   <%- if @listen_ip.is_a?(Array) then -%>
     <%- @listen_ip.each do |ip| -%>
-  listen       <%= ip %>:<%= @ssl_port %> <% if @ssl_listen_option %>ssl<% end %><% if @http2 == 'on' %> http2<% end %><% if @spdy == 'on' %> spdy<% end %><% if @listen_options %> <%= @listen_options %><% end %>;
+  listen       <%= ip %>:<%= @ssl_port %> <% if @ssl_listen_option %>ssl<% end %><% if @spdy == 'on' %> spdy<% end %><% if @listen_options %> <%= @listen_options %><% end %>;
```
Comment (smortex) [inline templates/server/server_ssl_header.erb:7]: The "legacy" `http2` on the listen line should still be emitted when `$nginx::nginx_version` is lower than 1.25.1.

Comment (C24-AK) [reply templates/server/server_ssl_header.erb]: Done, check commit :)

**Comment by smortex** [comment]
That looks better, thanks!

Can you please combine your 40+ commits in a single one?

From your working directory:

```
git rebase -i origin/main # Interactively rewrite history
```

This will bring your editor listing all commits in your branch.  Keep the first line unchanged, and replace `pick` with `squash` for all subsequent lines.  Save and quit, your editor will pop again and allow you to combine all the commit messages in a single one.

Then push your updated branch:

```
git push -f            # Send the changes (-f is required because we re-wrote history)
```

**Comment by C24-AK** [comment]
Hey @smortex,

I did what you explained and it feels like all my changes are gone?
All tests are failing

**Comment by C24-AK** [comment]
@smortex anything else to do or am I finished here?

File: spec/defines/resource_server_spec.rb [hunk 2]
```diff
@@ -745,18 +765,6 @@
...
-              attr: 'http2',
-              value: 'on',
-              match: %r{\s+listen\s+\*:443 ssl http2;}
-            },
-            {
-              title: 'should not set HTTP2',
-              attr: 'http2',
-              value: 'off',
-              match: %r{\s+listen\s+\*:443 ssl;}
-            },
```
Comment (smortex) [inline spec/defines/resource_server_spec.rb:759]: Should we not keep these tests for now?

**Comment by smortex** [comment]
> @smortex anything else to do or am I finished here?

That look good, the failures seems unrelated, I tried to run the test again, maybe it will pass and if not maybe this will ring a bell to some other contributor?

I added an in-line note about old test removal, and I wonder if we can keep them for now?

File: spec/defines/resource_server_spec.rb (hunk 2 above)
Comment (C24-AK) [reply spec/defines/resource_server_spec.rb:759]: You are right, I'll include them again

Comment (C24-AK) [reply spec/defines/resource_server_spec.rb]: Done.

**Comment by C24-AK** [comment]
@smortex I resolved your comment and added the tests

**Comment by kenyon** [comment]
I just reran the failing Puppet 7 job to see if it's consistent.

**Comment by kenyon** [comment]
I'm not sure how this change would be causing this error, but it seems like it is.

From https://github.com/voxpupuli/puppet-nginx/actions/runs/6809552993/job/18546775611?pr=1579#step:5:299

```
An error occurred in an `after(:suite)` hook.
Failure/Error: Parser.new(source, **(opts||{})).parse

JSON::ParserError:
  unexpected token at ''
# ./vendor/bundle/ruby/2.7.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
# ./vendor/bundle/ruby/2.7.0/gems/json-2.6.3/lib/json/common.rb:216:in `parse'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:99:in `load_filters'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:80:in `block in merge_filters'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:79:in `each'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:79:in `merge_filters'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:201:in `run_report'
# ./vendor/bundle/ruby/2.7.0/gems/rspec-puppet-4.0.0/lib/rspec-puppet/coverage.rb:186:in `report!'
# ./vendor/bundle/ruby/2.7.0/gems/voxpupuli-test-7.0.2/lib/voxpupuli/test/spec_helper.rb:21:in `block (2 levels) in <top (required)>'
```

**Comment by C24-AK** [comment]
@kenyon I am not very familiar with this. How would I fix this issue? I appreciate any help

**Comment by C24-AK** [comment]
Is this rdy to merge? @kenyon

**Pull request merged**